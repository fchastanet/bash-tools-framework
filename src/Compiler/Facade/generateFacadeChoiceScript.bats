#!/usr/bin/env bash

# shellcheck source=src/batsHeaders.sh
source "$(cd "${BATS_TEST_DIRNAME}/../.." && pwd)/batsHeaders.sh"

# shellcheck source=/src/Compiler/Facade/generateFacadeChoiceScript.sh
source "${srcDir}/Compiler/Facade/generateFacadeChoiceScript.sh"
# shellcheck source=/src/Compiler/Implement/mergeInterfacesFunctions.sh
source "${srcDir}/Compiler/Implement/mergeInterfacesFunctions.sh"
# shellcheck source=/src/Compiler/Implement/interfaceFunctions.sh
source "${srcDir}/Compiler/Implement/interfaceFunctions.sh"
# shellcheck source=/src/Compiler/Implement/assertInterface.sh
source "${srcDir}/Compiler/Implement/assertInterface.sh"
# shellcheck source=/src/Compiler/Implement/filter.sh
source "${srcDir}/Compiler/Implement/filter.sh"
# shellcheck source=/src/Compiler/Implement/parse.sh
source "${srcDir}/Compiler/Implement/parse.sh"
# shellcheck source=/src/Compiler/Implement/interfaces.sh
source "${srcDir}/Compiler/Implement/interfaces.sh"
# shellcheck source=/src/Compiler/Embed/getSrcDirsFromOptions.sh
source "${srcDir}/Compiler/Embed/getSrcDirsFromOptions.sh"
# shellcheck source=/src/Compiler/findFunctionInSrcDirs.sh
source "${srcDir}/Compiler/findFunctionInSrcDirs.sh"
# shellcheck source=/src/Filters/trimEmptyLines.sh
source "${srcDir}/Filters/trimEmptyLines.sh"
# shellcheck source=/src/Filters/removeExternalQuotes.sh
source "${srcDir}/Filters/removeExternalQuotes.sh"
# shellcheck source=/src/Assert/bashFrameworkFunction.sh
source "${srcDir}/Assert/bashFrameworkFunction.sh"
# shellcheck source=/src/Assert/posixFunctionName.sh
source "${srcDir}/Assert/posixFunctionName.sh"

function Compiler::Facade::generateFacadeChoiceScript::NoImplementDirective { #@test
  local scriptFile="${BATS_TEST_DIRNAME}/testsData/generateFacadeChoiceScript.noImplementDirective.sh"
  run Compiler::Facade::generateFacadeChoiceScript "${scriptFile}"
  assert_failure 2
  assert_lines_count 1
  assert_output --partial "WARN    - in ${scriptFile} - FACADE doesn't implement any interface"
}

function Compiler::Facade::generateFacadeChoiceScript::OneImplementDirective { #@test
  local -a _COMPILE_FILE_ARGUMENTS=(
    --src-dir "${BATS_TEST_DIRNAME}/testsData"
  )
  local scriptFile="${BATS_TEST_DIRNAME}/testsData/generateFacadeChoiceScript.oneImplementDirective.sh"
  local status=0
  export _COMPILE_FILE_ARGUMENTS
  Compiler::Facade::generateFacadeChoiceScript "${scriptFile}" >"${BATS_TEST_TMPDIR}/result" 2>&1 || status=$?
  [[ "${status}" = "0" ]]
  diff "${BATS_TEST_TMPDIR}/result" "${BATS_TEST_DIRNAME}/testsData/generateFacadeChoiceScript.oneImplementDirective.expected.txt" >&3 || return 1
}

function Compiler::Facade::generateFacadeChoiceScript::TwoImplementDirective { #@test
  local -a _COMPILE_FILE_ARGUMENTS=(
    --src-dir "${BATS_TEST_DIRNAME}/testsData"
  )
  local scriptFile="${BATS_TEST_DIRNAME}/testsData/generateFacadeChoiceScript.twoImplementDirective.sh"
  local status=0
  export _COMPILE_FILE_ARGUMENTS
  Compiler::Facade::generateFacadeChoiceScript "${scriptFile}" >"${BATS_TEST_TMPDIR}/result" 2>&1 || status=$?
  [[ "${status}" = "0" ]]
  diff "${BATS_TEST_TMPDIR}/result" "${BATS_TEST_DIRNAME}/testsData/generateFacadeChoiceScript.twoImplementDirective.expected.txt" >&3 || return 1
}
