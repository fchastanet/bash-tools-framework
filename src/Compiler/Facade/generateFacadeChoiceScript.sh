#!/usr/bin/env bash

# @description generate facade choice script
# - read all IMPLEMENT directives from original script file
# - deduce methods implemented by the script
# - generate a switch case that will call for arg $1,
#   the corresponding function of the interface
# - default case generates an error
# @example
#   # example of choice script generated
#   local action="$1"
#   shift || true
#   case ${action} in
#     interfaceFunction1)
#       interfaceFunction1 "$@"
#       ;;
#     interfaceFunction2)
#       interfaceFunction2 "$@"
#       ;;
#     *)
#       Log::displayError "invalid action requested: ${action}"
#       exit 1
#       ;;
#   esac
#   exit 0
#
# @arg $1 scriptFile:String the file from which a facade will be generated
# @exitcode 1 on template file not found or non readable file
# @exitcode 2 if FACADE doesn't implement any interface
# @stderr diagnostics information is displayed
# @see Compiler::Implement::mergeInterfacesFunctions
Compiler::Facade::generateFacadeChoiceScript() {
  local scriptFile="$1"

  local interfacesFunctionsStr
  interfacesFunctionsStr="$(Compiler::Implement::mergeInterfacesFunctions "${scriptFile}")" || return $?

  if [[ -z "${interfacesFunctionsStr}" ]]; then
    Log::displayWarning "in ${scriptFile} - FACADE doesn't implement any interface"
    return 2
  fi

  # generate code
  (
    printf $"local action="\$1"\n"
    printf 'shift || true\n'
    printf $"case \${action} in\n"
    # The function names have already been asserted
    local interfacesFunction
    while IFS='' read -r interfacesFunction; do
      printf "  %s)\n" "${interfacesFunction}"
      printf "    %s \"\$@\"\n" "${interfacesFunction}"
      printf "    ;;\n"
    done <<<"${interfacesFunctionsStr}"
    printf "  *)\n"
    printf '    if Assert::functionExists defaultFacadeAction; then\n'
    # shellcheck disable=SC2016
    printf '      defaultFacadeAction "$1" "$@"\n'
    printf '    else\n'
    printf $"      Log::displayError \"invalid action requested: \${action}\"\n"
    printf "      exit 1\n"
    printf '    fi\n'
    printf "    ;;\n"
    printf 'esac\n'
    printf 'exit 0\n'
  )

}
