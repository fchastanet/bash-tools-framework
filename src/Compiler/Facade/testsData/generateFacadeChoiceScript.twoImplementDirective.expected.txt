local action=$1
shift || true
case ${action} in
  install)
    install "$@"
    ;;
  breakOnConfigFailure)
    breakOnConfigFailure "$@"
    ;;
  breakOnTestFailure)
    breakOnTestFailure "$@"
    ;;
  *)
    if Assert::functionExists defaultFacadeAction; then
      defaultFacadeAction "$1" "$@"
    else
      Log::displayError "invalid action requested: ${action}"
      exit 1
    fi
    ;;
esac
exit 0
